### PR TITLE
Don't create channel role/rolebinding if the channel is in ACM NS

### PR DESCRIPTION
**Pull request by xiangjingli**
Signed-off-by: Xiangjing Li <xiangli@redhat.com>

https://github.com/open-cluster-management/backlog/issues/11921

The fix includes:

1. Don't create channel role/rolebinding if the channel is in ACM NS
2. Remove existing channel role/rolebinding if the channel is in ACM NS

### CODE REVIEW

File: pkg/controller/channel/channel_controller.go [hunk 1]
```diff
@@ -229,14 +229,26 @@ func (r *ReconcileChannel) Reconcile(request reconcile.Request) (reconcile.Resul
 		return reconcile.Result{}, nil
 	}
 
-	err = r.validateClusterRBAC(instance, log)
+	// find the channel controller pod namespace, it is running in the ACM namespece
+	podNamespace, err := utils.FindEnvVariable("POD_NAMESPACE")
+	if err != nil {
+		return reconcile.Result{}, gerr.Wrapf(err, "Failed to find ACM Namespace. err: %v", err)
```
Comment (rokej) [inline pkg/controller/channel/channel_controller.go:235]: Failed to find POD_NAMESPACE environment variable, not ACM Namespace?

File: pkg/controller/channel/channel_controller.go [hunk 2]
```diff
@@ -372,7 +384,13 @@ func (r *ReconcileChannel) syncReferredObjAnnotation(
 	return nil
 }
 
-func (r *ReconcileChannel) validateClusterRBAC(instance *chv1.Channel, logger logr.Logger) error {
+func (r *ReconcileChannel) validateClusterRBAC(instance *chv1.Channel, logger logr.Logger, podNamespace string) error {
+	if instance.Namespace == podNamespace {
+		logger.Info(fmt.Sprintf("Don't create role and rolebinding as the channel %v/%v is in the ACM Namespace %v",
```
Comment (rokej) [inline pkg/controller/channel/channel_controller.go:389]: channel is in the channel controller's namespace, not ACM namespace.

File: pkg/controller/channel/channel_controller.go [hunk 3]
```diff
@@ -487,3 +505,34 @@ func (r *ReconcileChannel) setupRole(instance *chv1.Channel, role *rbac.Role) er
 
 	return nil
 }
+
+// Clean up channel role/rolebinding if the channel is located in the ACM system Namespace,
+// so the ACM NameSpace Secrets won't be exposed to managed clusters.
+// The channels created in the ACM system NS are only used by hub standalone subscriptions.
+func (r *ReconcileChannel) cleanRoleFromAcmNS(instance *chv1.Channel, logger logr.Logger, podNamespace string) error {
+	if instance.Namespace != podNamespace {
+		logger.Info(fmt.Sprintf("The channel %v/%v is not in the ACM Namespace %v, skipping...",
```
Comment (rokej) [inline pkg/controller/channel/channel_controller.go:514]: Same. Channel is not in the channel controller's namespace?

**Comment by xiangjingli** [comment]
@rokej Thanks Roke! The previous fix assumes that the channel pod is running in the same ACM sysmte Namespace.  That may not be correct as users can install app pods in different namespace.

The new fix is to get the ACM system Namespace by finding the MultiClusterHub resource NS via the dynamic client. Please see if it is OK

**Review by rokej — COMMENTED** [review]
@xiangjingli Thanks, the changed code looks good. Have you thought about making use of the existing channel validation webhook to prevent users from creating a channel in ACM namespace? 

Another thought is .. is it enough to just document that channel creates these role and rolebinding to allow certain group to view secrets so that users build RBAC accordingly? A channel can be created in any namespace where there are secrets that users want to protect.

**Comment by xiangjingli** [comment]
@rokej  The trouble is we still need to allow creating channels in ACM system namespace. This is used by ACM installer for installing some helm charts.  

Now the channel namespace only allows users defined in the ACM managed cluster user group to read channel secrets. Normal end users won't have this permission. I think that is good enough to log it as a known issue and suggest end user to create channels in a new NS. Thus no existing secrets won't be exposed to the ACM managed cluster users

**Comment by rokej** [comment]
@xiangjingli I see. You are right about installation.

**Review by rokej — APPROVED** [review]

**Pull request merged**